### PR TITLE
Add full support for ES modules

### PR DESCRIPTION
**Pull request by guillaumervls**
Hey @benjie,

Here is some equivalent of what I did to Graphile Migrate but for Graphile Worker this time.

Again it's just a start / suggestion, because of the accelerating trend of moving from CommonJS to ESM. I guess there will be issues with Typescript e.g. https://github.com/microsoft/TypeScript/issues/43329 but I'm not fluent in Typescript...

Cheers !

### CODE REVIEW

**Comment by benjie** [comment]
I definitely plan to move everything to ESM long-term, so thanks for raising this and getting the discussion started :raised_hands:

**Comment by stephane-klein** [comment]
Does this Pull Request allow to solve #293 ?

File: src/module.ts [hunk 1]
```diff
@@ -10,6 +10,11 @@ function stripBOM(str: string) {
   return str;
 }
 
+export function fauxImport(spec: string) {
+  // Add a timestamp parameter to bust cache
+  return import(`${spec}?ts=${Date.now()}`);
```
[the remark this answers is not in the document]
Comment (stephane-klein) [reply src/module.ts:15]: > We'll need to solve that for this to be equivalent to current functionality.

@benjie Is this comment blocking the Pull Request?

**Comment by boromisp** [comment]
TLDR: use an external filesystem watcher, and restart the worker if a task executor has changed. Everything else is a pain.

---

The HMR story in node seems to be bit of a mess. The related chromium/v8 API is "not well" supported and has outstanding wontfix issues: https://bugs.chromium.org/p/chromium/issues/detail?id=1063378

I've seen some example implementations for cache busting ESM imports with loader hooks.  As I understand this, the current code is already doing something like this, just for CommonJS modules.

 - https://github.com/nodejs/modules/issues/307#issuecomment-764560656
 - https://dev.to/giltayar/mock-all-you-want-supporting-es-modules-in-the-testdouble-js-mocking-library-3gh1

The `vm.Module` (https://nodejs.org/api/vm.html#class-vmmodule) seems to be a lower level API, that might give more control than the loader hooks. I have zero experience with either of these.

 - https://gist.github.com/marxangels/68c1e50fe144c4ec4c911a6534aa7e32

---

A totally different approach would be spinning up worker threads for each task execution. That way the module cache would be thrown away after each execution. However that would limit the communication with the task executor, for example, you won't be able to pass a `pg` connection via `postMessage` (https://nodejs.org/api/worker_threads.html#portpostmessagevalue-transferlist)

 - https://github.com/nodejs/modules/issues/307#issuecomment-858729422

---

See also: https://github.com/mochajs/mocha/issues/4374

**Comment by benjie** [comment]
(But the restart would need to be soft because you don’t want to break running tasks because you edited an unrelated task file, so now every time you save you have to wait for your longest running existing task to complete before you can start processing new tasks again… or you could spin up the new worker earlier and just let the old worker exit on it’s own time, but…)

I’m considering removing watch mode from worker and encouraging people to use nodemon/etc instead.

**Comment by maxcountryman** [comment]
> I definitely plan to move everything to ESM long-term, so thanks for raising this and getting the discussion started 🙌

Has there been progress toward this?

**Comment by benjie** [comment]
> Has there been progress toward this?

No, the above issue needs solving and has yet to be solved, which is why this PR is still open. Progress towards this is tracked in this PR; to summarise the messages above your request for you: I'm not happy relying on an external restart mechanism for a job queue due to the need for soft restarts, and the ESM fauxImport is not a sufficient replacement for fauxRequire.

Should you need to use ESM with Graphile Worker you can do so using library mode (where you import the tasks and build the task list yourself, rather than having us scan the folder and do it for you); this has been supported for years.
